app/models/mixins/authentication_mixin.rb

```diff
@@ -129,24 +129,17 @@ def update_authentication(data, options = {})
 
     data.each_pair do |type, value|
       cred = authentication_type(type)
-      current = {:new => nil, :old => nil}
-
-      unless value.key?(:userid) && value[:userid].blank?
-        current[:new] = {:user => value[:userid], :password => value[:password], :auth_key => value[:auth_key]}
-      end
-      current[:old] = {:user => cred.userid, :password => cred.password, :auth_key => cred.auth_key} if cred
 
       # Raise an error if required fields are blank
       Array(options[:required]).each { |field| raise(ArgumentError, "#{field} is required") if value[field].blank? }
 
+      attributes = [:userid, :password, :auth_key]
       # If old and new are the same then there is nothing to do
-      next if current[:old] == current[:new]
+      next if attributes.all? { |a| value[a] == cred.try(a) }
 
       # Check if it is a delete
-      if value.key?(:userid) && value[:userid].blank?
-        current[:new] = nil
-        next if options[:save] == false
-        authentication_delete(type)
+      if required_credential_fields(type).any? { |field| value.key?(field) && value[field].blank? }
```

Comment:
If the key doesn't existing, it will return nil, which is covered by the blank check, so I think you can simplify to

```ruby
if required_credential_fields(type).any? { |field| value[field].blank? }
```